crates/nu-plugin/src/plugin/mod.rs

```diff
@@ -84,121 +108,79 @@ pub(crate) fn create_command(path: &Path, shell: Option<&Path>) -> CommandSys {
-    let program_name = plugin_cmd.get_program().to_os_string().into_string();
-
-    plugin_cmd.envs(current_envs);
-    let mut child = plugin_cmd.spawn().map_err(|err| {
-        let error_msg = match err.kind() {
-            ErrorKind::NotFound => match program_name {
-                Ok(prog_name) => {
-                    format!("Can't find {prog_name}, please make sure that {prog_name} is in PATH.")
-                }
-                _ => {
-                    format!("Error spawning child process: {err}")
-                }
-            },
-            _ => {
-                format!("Error spawning child process: {err}")
-            }
-        };
-
-        ShellError::PluginFailedToLoad { msg: error_msg }
-    })?;
-
-    let mut stdin_writer = child
+fn make_plugin_interface(
+    mut child: Child,
+    identity: Arc<PluginIdentity>,
+) -> Result<PluginInterface, ShellError> {
+    let stdin = child
         .stdin
         .take()
         .ok_or_else(|| ShellError::PluginFailedToLoad {
             msg: "plugin missing stdin writer".into(),
         })?;
-    let mut stdout_reader = child
+
+    let mut stdout = child
         .stdout
         .take()
         .ok_or_else(|| ShellError::PluginFailedToLoad {
-            msg: "Plugin missing stdout reader".into(),
+            msg: "Plugin missing stdout writer".into(),
         })?;
-    let encoding = get_plugin_encoding(&mut stdout_reader)?;
-
-    // Create message to plugin to indicate that signature is required and
-    // send call to plugin asking for signature
-    let encoding_clone = encoding.clone();
-    // If the child process fills its stdout buffer, it may end up waiting until the parent
-    // reads the stdout, and not be able to read stdin in the meantime, causing a deadlock.
-    // Writing from another thread ensures that stdout is being read at the same time, avoiding the problem.
-    std::thread::spawn(move || {
-        encoding_clone.encode_call(&PluginCall::Signature, &mut stdin_writer)
-    });
 
-    // deserialize response from plugin to extract the signature
-    let reader = stdout_reader;
-    let mut buf_read = BufReader::with_capacity(OUTPUT_BUFFER_SIZE, reader);
-    let response = encoding.decode_response(&mut buf_read)?;
-
-    let signatures = match response {
-        PluginResponse::Signature(sign) => Ok(sign),
-        PluginResponse::Error(err) => Err(err.into()),
-        _ => Err(ShellError::PluginFailedToLoad {
-            msg: "Plugin missing signature".into(),
-        }),
-    }?;
-
-    match child.wait() {
-        Ok(_) => Ok(signatures),
-        Err(err) => Err(ShellError::PluginFailedToLoad {
-            msg: format!("{err}"),
-        }),
-    }
+    let encoder = get_plugin_encoding(&mut stdout)?;
+
+    let reader = BufReader::with_capacity(OUTPUT_BUFFER_SIZE, stdout);
+
+    let mut manager = PluginInterfaceManager::new(identity, (Mutex::new(stdin), encoder.clone()));
```

Comment:
1. okay, it's now `Copy`
2. The `Mutex` is required when writing to `ChildStdin` because there is no inherent lock there, and to write to it requires a `&mut` reference. We want all of the writes to be atomic and it's not possible to get the mutable reference anyway. We don't need it for `std::io::Stdout` though because of the inherent lock available.